include/substrait/type/Type.h

```diff
@@ -0,0 +1,701 @@
+};
+
+template <>
+struct TypeTraits<TypeKind::kFixedBinary> {
+  static constexpr const char* signature = "fbin";
+  static constexpr const char* typeString = "fixedbinary";
+};
+
+template <>
+struct TypeTraits<TypeKind::kDecimal> {
+  static constexpr const char* signature = "dec";
+  static constexpr const char* typeString = "decimal";
+};
+
+template <>
+struct TypeTraits<TypeKind::kStruct> {
+  static constexpr const char* signature = "struct";
+  static constexpr const char* typeString = "struct";
+};
+
+template <>
+struct TypeTraits<TypeKind::kList> {
+  static constexpr const char* signature = "list";
+  static constexpr const char* typeString = "list";
+};
+
+template <>
+struct TypeTraits<TypeKind::kMap> {
+  static constexpr const char* signature = "map";
+  static constexpr const char* typeString = "map";
+};
+
+template <>
+struct TypeTraits<TypeKind::kUserDefined> {
+  static constexpr const char* signature = "u!name";
+  static constexpr const char* typeString = "user defined type";
+};
+
+class ParameterizedType {
+ public:
+  explicit ParameterizedType(bool nullable = false) : nullable_(nullable) {}
+
+  [[nodiscard]] virtual std::string signature() const = 0;
+
+  [[nodiscard]] virtual TypeKind kind() const = 0;
+
+  /// Deserialize substrait raw type string into Substrait extension  type.
+  /// @param rawType - substrait extension raw string type
+  static std::shared_ptr<const ParameterizedType> decode(
+      const std::string& rawType);
+
+  [[nodiscard]] const bool& nullable() const {
+    return nullable_;
+  }
+
+  [[nodiscard]] bool nullMatch(
+      const std::shared_ptr<const ParameterizedType>& type) const {
+    return nullable() || nullable() == type->nullable();
+  }
+  /// Test type is a Wildcard type or not.
+  [[nodiscard]] virtual bool isWildcard() const {
+    return false;
+  }
+
+  [[nodiscard]] virtual bool isMatch(
+      const std::shared_ptr<const ParameterizedType>& type) const = 0;
+
+ private:
+  const bool nullable_;
+};
+
+using ParameterizedTypePtr = std::shared_ptr<const ParameterizedType>;
+
+class Type : public ParameterizedType {
+ public:
+  explicit Type(bool nullable = false) : ParameterizedType(nullable) {}
+};
+
+using TypePtr = std::shared_ptr<const Type>;
+
+/// Types used in function argument declarations.
+template <TypeKind Kind>
+class TypeBase : public Type {
+ public:
+  explicit TypeBase(bool nullable = false) : Type(nullable) {}
+
+  [[nodiscard]] std::string signature() const override {
+    return TypeTraits<Kind>::signature;
+  }
+
+  [[nodiscard]] TypeKind kind() const override {
+    return Kind;
+  }
+
+  [[nodiscard]] bool isMatch(
+      const std::shared_ptr<const ParameterizedType>& type) const override {
+    return kind() == type->kind() && nullMatch(type);
+  }
+};
+
+template <TypeKind Kind>
+class ScalarType : public TypeBase<Kind> {
+ public:
+  explicit ScalarType(bool nullable) : TypeBase<Kind>(nullable) {}
+};
```

Comment:
Same question as earlier.  This class does not appear to add anything to `TypeBase`.

Reply:
This class intend to be  a placeHolder for scalar type.